api/src/wfl/source.clj

```diff
@@ -285,13 +287,33 @@
            (run! #(write-snapshot-id source %)))
       (log/debug (format "%s Running snapshot jobs updated." (log-prefix source))))))
 
+;; Workloads in general may be updated more frequently,
+;; but overpolling the TDR increases the chances of:
+;; - creating many single-row / low-cardinality snapshots
+;; - locking the dataset / running into dataset locks
+(def ^:private tdr-source-polling-interval-minutes 20)
+
+(defn ^:private tdr-source-should-poll?
+  "Return true if it's been at least `tdr-source-polling-interval-minutes`
+   since `last_checked` -- when we last checked for new rows in the TDR."
+  [{:keys [type id last_checked] :as _source} utc-now]
+  (let [checked            (timestamp-to-offsetdatetime last_checked)
+        minutes-since-poll (.between ChronoUnit/MINUTES checked utc-now)]
+    (log/debug {:type               type
+                :id                 id
+                :minutes-since-poll minutes-since-poll
+                :polling-interval   tdr-source-polling-interval-minutes})
+    (<= tdr-source-polling-interval-minutes minutes-since-poll)))
+
 (defn ^:private update-tdr-source
   "Check for new data in TDR from `source`, create new snapshots,
   insert resulting job creation ids into database and update the
   timestamp for next time."
   [{:keys [stopped] :as source}]
-  (when-not stopped
-    (find-and-snapshot-new-rows source (utc-now)))
+  (let [now          (utc-now)
+        should-poll? (tdr-source-should-poll? source now)]
+    (when (and (not stopped) should-poll?)
+      (find-and-snapshot-new-rows source now)))
```

Comment:
Inline `should-poll?` so it doesn't test a `stopped` workload.

Reply:
Thanks.  (I thought I'd get short-circuiting but that doesn't work if you've already evaluated the expression!)